src/Columns/ColumnVector.cpp

```diff
@@ -471,6 +477,115 @@ void ColumnVector<T>::insertRangeFrom(const IColumn & src, size_t start, size_t
     memcpy(data.data() + old_size, &src_vec.data[start], length * sizeof(data[0]));
 }
 
+static inline UInt64 blsr(UInt64 mask)
+{
+#ifdef __BMI__
+    return _blsr_u64(mask);
+#else
+    return mask & (mask-1);
+#endif
+}
+
+DECLARE_DEFAULT_CODE(
+template <typename T, typename Container, size_t SIMD_BYTES>
+inline void doFilterAligned(const UInt8 *& filt_pos, const UInt8 *& filt_end_aligned, const T *& data_pos, Container & res_data)
+{
+    while (filt_pos < filt_end_aligned)
+    {
+        UInt64 mask = bytes64MaskToBits64Mask(filt_pos);
+
+        if (0xffffffffffffffff == mask)
+        {
+            res_data.insert(data_pos, data_pos + SIMD_BYTES);
+        }
+        else
+        {
+            while (mask)
+            {
+                size_t index = std::countr_zero(mask);
+                res_data.push_back(data_pos[index]);
+                mask = blsr(mask);
+            }
+        }
+
+        filt_pos += SIMD_BYTES;
+        data_pos += SIMD_BYTES;
+    }
+}
+)
+
+DECLARE_AVX512VBMI2_SPECIFIC_CODE(
+template <size_t ELEMENT_WIDTH>
+inline void compressStoreAVX512(const void *src, void *dst, const UInt64 mask)
+{
+    __m512i vsrc = _mm512_loadu_si512(src);
+    if constexpr (ELEMENT_WIDTH == 1)
+        _mm512_mask_compressstoreu_epi8(dst, static_cast<__mmask64>(mask), vsrc);
+    else if constexpr (ELEMENT_WIDTH == 2)
+        _mm512_mask_compressstoreu_epi16(dst, static_cast<__mmask32>(mask), vsrc);
+    else if constexpr (ELEMENT_WIDTH == 4)
+        _mm512_mask_compressstoreu_epi32(dst, static_cast<__mmask16>(mask), vsrc);
+    else if constexpr (ELEMENT_WIDTH == 8)
+        _mm512_mask_compressstoreu_epi64(dst, static_cast<__mmask8>(mask), vsrc);
+}
+
+template <typename T, typename Container, size_t SIMD_BYTES>
+inline void doFilterAligned(const UInt8 *& filt_pos, const UInt8 *& filt_end_aligned, const T *& data_pos, Container & res_data)
+{
+    static constexpr size_t VEC_LEN = 64;   /// AVX512 vector length - 64 bytes
+    static constexpr size_t ELEMENT_WIDTH = sizeof(T);
+    static constexpr size_t ELEMENTS_PER_VEC = VEC_LEN / ELEMENT_WIDTH;
+    static constexpr UInt64 KMASK = 0xffffffffffffffff >> (64 - ELEMENTS_PER_VEC);
+
+    size_t current_offset = res_data.size();
+    size_t reserve_size = res_data.size();
+    size_t alloc_size = SIMD_BYTES * 2;
+
+    while (filt_pos < filt_end_aligned)
+    {
+        /// to avoid calling resize too frequently, resize to reserve buffer.
+        if (reserve_size - current_offset < SIMD_BYTES)
+        {
+            reserve_size += alloc_size;
+            res_data.resize(reserve_size);
+            alloc_size *= 2;
+        }
```

Comment:
@alexey-milovidov @rschu1ze 
I am able to reproduce https://github.com/ClickHouse/ClickHouse/issues/39895 if built and unittest with **msan**.

The problem is happen here: we reserve a buffer in advance and then use the pointer to write data with AVX512 instruction. *msan* complains for such behaviour.
If I replace Line 550 with explicitly filling to touch memory, the problem gone: 
```cpp
res_data.resize_fill(reserve_size, static_cast<T>(0));
```
Should I disable it in `MEMORY_SANITIZER` mode like [LZ4_decompress_faster.cpp#L279](https://github.com/ClickHouse/ClickHouse/blob/master/src/Compression/LZ4_decompress_faster.cpp#L279)?

Reply:
For completeness, I'll paste my minimal repro, but you were obviously quicker 😄 

Explicitly initialization of allocated memory via `resize_fill` adds unnecessary performance overhead. So yes, I would favor `!defined(MEMORY_SANITIZER)` like the file you mentioned. What about something like

```
#if defined(MEMORY_SANITIZER)
    res_data.resize_fill(reserve_size, static_cast<T>(0)); // MSan doesn't recognize that all allocated memory is written by AVX-512 intrinsics.
#else
    res_data.resize(reserve_size);
#endif
```

---

```
TEST(ColumnSparse, Filter)
{
    const size_t rows = 1000;

    auto col_src = ColumnVector<UInt64>::create();
    for (size_t i = 0; i < rows; ++i)
        col_src->getData().push_back(1);

    PaddedPODArray<UInt8> filter(rows);
    for (size_t i = 0; i < rows; ++i)
     filter[i] = i % 2 == 0;

    auto col_dst = col_src->filter(filter, -1);

    if (col_dst->compareAt(0, 0, *col_dst, 0) != 0)
    {
        throw Exception(error_code, "Columns are unequal");
    }
}
```

Reply:
It sounds good so we can still tests on other parts.
```cpp
#if defined(MEMORY_SANITIZER)
    res_data.resize_fill(reserve_size, static_cast<T>(0)); // MSan doesn't recognize that all allocated memory is written by AVX-512 intrinsics.
#else
    res_data.resize(reserve_size);
#endif
```

But the compiler will raise an warning `-Wembedded-directive` to `#if defined(...)` since we are declaring the function within marco arguments. We may need to make it as inline function to avoid warning (like ColumnVector.cpp#L480 in this PR).

Reply:
Sounds good. Let's make that one-liner an inlineable function.